torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -80,7 +80,19 @@ bool FunctionParameter::check(PyObject* obj) {
     case ParameterType::TENSOR: {
       return THPVariable_Check(obj);
     }
-    case ParameterType::SCALAR: return THPUtils_checkDouble(obj);
+    case ParameterType::SCALAR: {
+      // NOTE: we don't currently accept most NumPy types as Scalars. np.float64
+      // is okay because it's a subclass of PyFloat. We may want to change this
+      // in the future.
+      if (THPUtils_checkDouble(obj)) {
+        return true;
+      }
+      if (THPVariable_Check(obj)) {
+        auto& var = ((THPVariable*)obj)->cdata;
+        return !var.requires_grad() && var.dim() == 0;
```

Comment:
Thinking about this more, I feel like this binding-vs-not based on requires_grad is potentially confusing.  Are there cases where if we don't have the requires_grad check here we would select the "wrong" function (i.e. there is a proper tensor overload before the scalar overload in the signature list)?  Because if not, wouldn't it be nicer to add the requires_grad check to scalarWithDefault; i.e. something like "bound to function [signature], but argument [Y] is a tensor that requires_grad, which isn't supported with Scalar parameters".

Reply:
At that point it's too late. For example we'll select `add(Tensor, Scalar)` instead of `add(Tensor, Tensor)`. The second works and is differentiable (although potentially a slightly slower implementation).

The few cases where you get a TypeError are not so bad I think:

```
>>> torch.addmm(t1, t2, t3, alpha=torch.variable(2, requires_grad=True))

TypeError: addmm(): argument 'alpha' must be float, not Variable
```

 We should get rid of Scalars as separate overloads, too avoid this problem, but after we finish merging Variable and Tensor.

Reply:
You are right the error message is good and I agree with the longer term solution of getting rid of Scalar in the interface.  The "picking the wrong overload" does seem a little strange though -- couldn't we just reorder the (few) cases in Declarations.cwrap where this is a problem so we prefer the tensor overload?

Reply:
(cont.) with the idea that we want to get rid of the scalar overloads anyway, so putting them at the bottom seems right anyway.